pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,440 @@
+}
+
+// GetInt returns the scalar as a int, or an error otherwise
+func (n *leafNodeImpl) GetInt() (int, error) {
+	switch n.val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return int(n.val.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return int(n.val.Uint()), nil
+	case reflect.Float32, reflect.Float64:
+		return int(n.val.Float()), nil
+	}
+	return 0, newConversionError(n.val, "int")
+}
+
+// GetFloat returns the scalar as a float64, or an error otherwise
+func (n *leafNodeImpl) GetFloat() (float64, error) {
+	switch n.val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return float64(n.val.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return float64(n.val.Uint()), nil
+	case reflect.Float32, reflect.Float64:
+		return float64(n.val.Float()), nil
+	}
+	return 0, newConversionError(n.val, "float")
+}
+
+// GetString returns the scalar as a string, or an error otherwise
+func (n *leafNodeImpl) GetString() (string, error) {
+	if n.val.Kind() == reflect.String {
+		return n.val.String(), nil
+	}
+	return "", newConversionError(n.val, "string")
+}
+
+// convert a string to a bool using standard yaml constants
+func convertToBool(text string) (bool, error) {
+	lower := strings.ToLower(text)
+	if lower == "y" || lower == "yes" || lower == "on" || lower == "true" {
+		return true, nil
+	} else if lower == "n" || lower == "no" || lower == "off" || lower == "false" {
+		return false, nil
+	}
+	return false, newConversionError(reflect.ValueOf(text), "bool")
+}
+
+func fieldNameToKey(field reflect.StructField) string {
+	name := field.Name
+	if tagtext := field.Tag.Get("yaml"); tagtext != "" {
+		name = tagtext
+	} else if tagtext := field.Tag.Get("json"); tagtext != "" {
+		name = tagtext
+	} else if tagtext := field.Tag.Get("mapstructure"); tagtext != "" {
+		name = tagtext
+	}
+	// skip any additional specifiers such as ",omitempty"
+	if commaPos := strings.IndexRune(name, ','); commaPos != -1 {
+		name = name[:commaPos]
+	}
+	return name
+}
+
+func copyStruct(target reflect.Value, source node) error {
+	targetType := target.Type()
+	for i := 0; i < targetType.NumField(); i++ {
+		f := targetType.Field(i)
+		ch, _ := utf8.DecodeRuneInString(f.Name)
+		if unicode.IsLower(ch) {
+			continue
+		}
+		child, err := source.GetChild(fieldNameToKey(f))
+		if err == errNotFound {
+			continue
+		} else if err != nil {
+			return err
+		}
+		err = copyAny(target.FieldByName(f.Name), child)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copyMap(target reflect.Value, source node) error {
+	// TODO: Should handle maps with more complex types in a future PR
+	ktype := reflect.TypeOf("")
+	vtype := reflect.TypeOf("")
+	mtype := reflect.MapOf(ktype, vtype)
+	results := reflect.MakeMap(mtype)
+
+	mapKeys, err := source.ChildrenKeys()
```

Comment:
This would mean we could Unmarshal a struct into a map, no ? Should we prevent this and check that the source is also a map ?

Similar question for the other `copy*` func

Reply:
I think this is okay? It matches how the standard Unmarshal family of functions work. We'll keep an eye on it in future PRs, as this new function is used in more places, to see if we should decide to disable this behavior.